File: LongestConseutiveSubsequence/LongestConsecutiveSubsequence.cpp
/*
https://www.geeksforgeeks.org/problems/longest-consecutive-subsequence2449/1
*/

//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
class Solution {
  public:

    // Function to return length of longest subsequence of consecutive integers.
    int longestConsecutive(vector<int>& arr) {
        set<int>all;
        for(auto &vals: arr){
            all.insert(vals);
        }
        vector<int>newVec;
        for(auto &vals: all){
            newVec.push_back(vals);
        }
        int sz=newVec.size(), res=INT_MIN;
        for(int i=0; i<sz; ++i){
            int curr=newVec[i];
            if(i+1<=sz-1){
                int curidx=i+1, curlength=1;
                while(true){
                	if(curidx>=sz) break;
                    if(newVec[curidx]!=curr+1) break;
                    curr=newVec[curidx];
                    curlength++, curidx++;
                }
                res=max(res, curlength);
            }
        }
        return res;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;

        // Read first array
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }

        Solution ob;
        int res = ob.longestConsecutive(arr);

        cout << res << endl << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends

File: LongestConseutiveSubsequence/LongestConsecutiveSubsequence.java
/*
https://www.geeksforgeeks.org/problems/longest-consecutive-subsequence2449/1
*/

//{ Driver Code Starts
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            // Read first array
            String input = br.readLine();
            String[] inputArray = input.split(" ");
            int[] arr = Arrays.stream(inputArray).mapToInt(Integer::parseInt).toArray();

            Solution ob = new Solution();
            int res = ob.longestConsecutive(arr);

            System.out.println(res);
            System.out.println("~");
        }
    }
}

// } Driver Code Ends


class Solution {

    // Function to return length of longest subsequence of consecutive integers.
    public int longestConsecutive(int[] arr) {
        TreeSet<Integer>allValues=new TreeSet<>();
        for(int values: arr){
            allValues.add(values);
        }
        ArrayList<Integer>listAllValues=new ArrayList<>();
        Iterator itr=allValues.iterator();
        while(itr.hasNext()){
            int currentValue=(int)itr.next();
            listAllValues.add(currentValue);
        }
        int lengthOfList=listAllValues.size(), resultantLength=Integer.MIN_VALUE;
        for(int i=0; i<lengthOfList; ++i){
            int currentValue=listAllValues.get(i);
            if(i+1<=lengthOfList-1){
                int nextIndex=i+1, currentLength=1;
                while(true){
                    if(nextIndex>=lengthOfList) break;
                    if(listAllValues.get(nextIndex)!=currentValue+1) break;
                    currentValue=listAllValues.get(nextIndex);
                    currentLength++;
                    nextIndex++;
                }
                resultantLength=Math.max(resultantLength, currentLength);
            }
        }
        return resultantLength;
    }
}